### ISSUE TITLE
[Question] Why does the example use `likelihood` again even though `model` already calls `likelihood` internally?

### ISSUE THREAD

**Issue by daizhirui** [cornellius-gp/gpytorch#2188]
Hi, I am new to GPytorch. When I am looking at the [simple regression example](https://docs.gpytorch.ai/en/latest/examples/01_Exact_GPs/Simple_GP_Regression.html), I tried to reproduce the posterior covariance without GPytorch and compared the result with the one computed by the code from the regression example. The result is very different.

In the example, we have the following instances

```python
likelihood = gpytorch.likelihoods.GaussianLikelihood()
model = ExactGPModel(train_x, train_y, likelihood)
mll = gpytorch.mlls.ExactMarginalLogLikelihood(likelihood, model)
```

`model` and `mll` are used when we do the training:

```python
# Find optimal model hyperparameters
model.train()
likelihood.train()

# Use the adam optimizer
optimizer = torch.optim.Adam(model.parameters(), lr=0.1)  # Includes GaussianLikelihood parameters

# "Loss" for GPs - the marginal log likelihood
mll = gpytorch.mlls.ExactMarginalLogLikelihood(likelihood, model)

for i in range(training_iter):
    # Zero gradients from previous iteration
    optimizer.zero_grad()
    # Output from model
    output = model(train_x)
    # Calc loss and backprop gradients
    loss = -mll(output, train_y)
    loss.backward()
    print('Iter %d/%d - Loss: %.3f   lengthscale: %.3f   noise: %.3f' % (
        i + 1, training_iter, loss.item(),
        model.covar_module.base_kernel.lengthscale.item(),
        model.likelihood.noise.item()
    ))
    optimizer.step()
```

I check the source code of [ExactGP](https://github.com/cornellius-gp/gpytorch/blob/23b068b5d4ceb9f2822b8e3e6215ee357c5dd535/gpytorch/models/exact_gp.py#L244). It shows that the following `forward` method is called almost immediately by `ExactGP.__call__` when it is training:

```python
class ExactGPModel(gpytorch.models.ExactGP):
    def __init__(self, train_x, train_y, likelihood):
        super(ExactGPModel, self).__init__(train_x, train_y, likelihood)
        self.mean_module = gpytorch.means.ConstantMean()
        self.covar_module = gpytorch.kernels.ScaleKernel(gpytorch.kernels.RBFKernel())

    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return gpytorch.distributions.MultivariateNormal(mean_x, covar_x)
```

obviously, `likelihood` is not involved at `output = model(train_x)`. `likelihood` is used when `loss=-mll(output, train_y)` is executed.

This process is exactly the same as the standard GP regression. Now let's look at the prediction in this tutorial. The tutorial shows two different prediction procedures. One is showed in a markdown cell of the jupyter notebook,

```python
f_preds = model(test_x)
y_preds = likelihood(model(test_x))

f_mean = f_preds.mean
f_var = f_preds.variance
f_covar = f_preds.covariance_matrix
f_samples = f_preds.sample(sample_shape=torch.Size(1000,))
```

the other one is in a code cell of the notebook

```python
# Get into evaluation (predictive posterior) mode
model.eval()
likelihood.eval()

# Test points are regularly spaced along [0,1]
# Make predictions by feeding model through likelihood
with torch.no_grad(), gpytorch.settings.fast_pred_var():
    test_x = torch.linspace(0, 1, 51)
    observed_pred = likelihood(model(test_x))
```

If you look at the source code of [ExactGP](https://github.com/cornellius-gp/gpytorch/blob/23b068b5d4ceb9f2822b8e3e6215ee357c5dd535/gpytorch/models/exact_gp.py#L244), you will find that `likelihood` is used already when `model(test_x)` is executed!! So, **the question is: what is the purpose of calling `likelihood(model_output)` explicitly again?** I think it makes sense to call `f_preds.sample(sample_shape=torch.Size(1000,))` to sample the Gaussian distribution for `test_x` as the `noise` parameter from `likelihood` is already included in `f_preds`.

If we call `likelihood(model_output)` during prediction, we actually use the same Gaussian noise twice at two different locations:

at training:

$$
y(x) = f(x) + \epsilon, \epsilon \sim \mathcal{N}(0, \sigma^2)
$$

and at prediction:

$$
y(x^\*) = f(x^\* | X, y) + \epsilon, \epsilon \sim \mathcal{N}(0, \sigma^2)
$$

I think this is wrong. Please point it out if I miss something.

The following code shows my trial of reproducing the posterior variance:

```python
model.eval()
likelihood.eval()

with torch.no_grad():
    with gpytorch.settings.fast_pred_var():
        xs = torch.linspace(0, 1, 10)
        # compute the variance by GPytorch
        model_output = model(xs)
        var1 = model_output.variance.numpy()
        # compute the variance manually
        covar_train = model.covar_module(train_x).numpy()
        covar_inv = np.linalg.inv(covar_train + likelihood.noise.numpy() * np.eye(covar_train.shape[0]))
        covar_test_train = model.covar_module(xs, train_x).numpy()
        covar_test = model.covar_module(xs, xs).numpy()
        var2 = np.diag(covar_test) - ((covar_test_train @ covar_inv) * covar_test_train).sum(axis=1)
        diff = np.abs(var1 - var2)
        print(f"mean={diff.mean()}\nmin={diff.min()}\nmax={diff.max()}")

# example output:
# mean=9.467951975813265e-13
# min=2.639323492831941e-14
# max=1.7383046906202168e-12
```

and

```python
model.eval()
likelihood.eval()

with torch.no_grad():
    with gpytorch.settings.fast_pred_var():
        xs = torch.linspace(0, 1, 10)
        # compute the variance by GPytorch
        model_output = model(xs)
        var1 = likelihood(model_output).variance.numpy()  # <--------- the only difference
        # compute the variance manually
        covar_train = model.covar_module(train_x).numpy()
        covar_inv = np.linalg.inv(covar_train + likelihood.noise.numpy() * np.eye(covar_train.shape[0]))
        covar_test_train = model.covar_module(xs, train_x).numpy()
        covar_test = model.covar_module(xs, xs).numpy()
        var2 = np.diag(covar_test) - ((covar_test_train @ covar_inv) * covar_test_train).sum(axis=1)
        diff = np.abs(var1 - var2)
        print(f"mean={diff.mean()}\nmin={diff.min()}\nmax={diff.max()}")

# example output:
# mean=0.00015961483064945255
# min=0.00015961482707231087
# max=0.0001596148355272341
```

I think we should update the prediction part in the tutorial to be
```python
# Get into evaluation (predictive posterior) mode
model.eval()
likelihood.eval()

# Test points are regularly spaced along [0,1]
# Make predictions by feeding model through likelihood
with torch.no_grad(), gpytorch.settings.fast_pred_var():
    test_x = torch.linspace(0, 1, 51)
    observed_pred = model(test_x)  # <---------- remove the use of likelihood
```

**Comment by gpleiss** [comment]
> If you look at the source code of [ExactGP](https://github.com/cornellius-gp/gpytorch/blob/23b068b5d4ceb9f2822b8e3e6215ee357c5dd535/gpytorch/models/exact_gp.py#L244), you will find that likelihood is used already when model(test_x) is executed!!

It is called on the training data. (The line you are referring to produces the $\sigma^2 \boldsymbol I$ term of the $(\boldsymbol K + \sigma^2 \boldsymbol I)^{-1}$ term found in the posterior equations.

> So, the question is: what is the purpose of calling likelihood(model_output) explicitly again?

- Calling `model(test_x)` returns the posterior distribution of the latent function (i.e. $f(\boldsymbol x^*) \mid \mathcal D$, where $\mathcal D$ is the training data).
- Calling `likelihood(model(test_x))` returns the **predictive posterior** distribution (i.e. $y^* (\boldsymbol x^*) \mid \mathcal D$.

**Comment by daizhirui** [comment]
Thank you for pointing out the difference! I understand it now.
